### ISSUE TITLE
On Windows, `getcontext` returns an error

### ISSUE THREAD

**Issue by dinosaure** [dinosaure/esperanto#11]
It seems that `getcontext` returns a error and it's impossible for `pth` to spawn a new thread. Due to this error, a SEGFAULT is reached at one point where no threads was launched but `pth_key_setdata` expects a current one. I will try to find if the error come from Cosmopolitan or if it's a real barrier.

**Comment by dinosaure** [comment]
It seems not possible to use `pth` on Windows (it seems that only `w32` is supported). More concretely, Cosmopolitan does not implement
- `mcsc` machine context method (`makecontext(3)`/`setcontext(3)`) for Windows (`ESYS` is returned)
- `sas` stack setup method (`sigaltstack(2)`)
- `ss` stack setup method (`sigstack(2)`)

`sjljw32` is available but it works only for a 32-bits architecture which is too restrictive for us. We have 2 ways to solve this issue:
- implement required functions `mcsc`, `sas` or `ss` into Cosmopolitan
- modify `pth` to provide a Windows implementation of `pth_mctx` - we probably need to use the `mcsc` compilation path and check if we are into `IsWindows()` and do something else than `makecontext`/`setcontext`

**Comment by dinosaure** [comment]
The required interface of `pth_mctx` is:
```c
typedef struct pth_mctx_t pth_mctx_t;

pth_mctx_switch(pth_mctx_t* , pth_mctx_t *);
pth_mctx_restore(pth_mctx_t *);
pth_mctx_set(pth_mctx_t *, void (*f)(void), char *addr_lo, char *addr_hi);
```

**Comment by dinosaure** [comment]
The best possibily is to integrate [`winpthreads`](https://github.com/ldx/winpthreads) into `pth` and use `IsWindows()` to switch from one implementation to another depending on the platform (thx @dra27 for his help). OCaml multicore will use it too so we ensure a _certain_ coherence between Esperanto and OCaml Multicore.

**Comment by jart** [comment]
@dinosaure Cosmopolitan Libc has pthreads support now. https://github.com/jart/cosmopolitan/releases/tag/2.1 If there's an issue with setcontext() and getcontext() that are causing issues for fibers, then let me know and I'll fix it.

**Comment by dinosaure** [comment]
Thanks @jart, I will probably focus on that in 2~3 weeks :+1: - It's currently on top of my TODO list.

**Comment by jart** [comment]
Awesome to hear! Looking forward to seeing where it goes.

**Comment by dinosaure** [comment]
FIxed by #28 which uses Cosmopolitan 2.2 (which has `pthread_` functions).
